File: types/recipe.ts
// alche:me — Recipe Type Definitions

/** 代用テク情報（構造化） */
export interface SubstitutionInfo {
  original_name: string;
  original_brand?: string;
  reason: string;
  tips: string[];
  search_url?: string;
}

/** レシピステップ */
export interface RecipeStep {
  step: number;
  area: string;
  item_id: string;
  item_name: string;
  instruction: string;
  substitution_note?: string;
  substitution?: SubstitutionInfo;
  color_code?: string;
  color_name?: string;
  brand?: string;
  image_url?: string;
}

/** レシピ — Firestore: users/{userId}/recipes/{recipeId}
 *  テーマのみエントリ (source: "theme") もこのコレクションに統合。
 */
export interface Recipe {
  id: string;
  recipe_name: string;

  // --- テーマフィールド ---
  theme_title?: string;
  theme_description?: string;
  style_keywords?: string[];
  theme_status?: "active" | "liked" | "skipped";
  theme_context?: {
    weather?: string;
    occasion?: string;
    season?: string;
  };

  // --- レシピフィールド ---
  user_request: string;
  match_score?: number;
  thinking_process: string[];
  steps: RecipeStep[];
  context?: {
    weather?: string;
    occasion?: string;
    note?: string;
  };
  pro_tips: string[];
  preview_image_url?: string;
  character_theme?: "cute" | "cool" | "elegant";
  is_favorite: boolean;
  source?: "manual" | "ai" | "theme";
  feedback?: {
    user_rating: "liked" | "neutral" | "disliked";
    created_at: string;
  };
  created_at: string;
}

/** レシピ生成結果（Alchemist Agent → Next.js API） */
export interface RecipeResult {
  success: boolean;
  recipe?: Recipe;
  thinking_process?: string[];
  error?: string;
}

File: agent/alcheme/schemas/recipe.py
from pydantic import BaseModel, Field


class RecipeStep(BaseModel):
    step: int
    area: str
    item_id: str
    item_name: str
    technique: str = "standard"  # "standard" | "substitution" | "layering"
    instruction: str
    substitution_note: str | None = None


class RecipeValidation(BaseModel):
    all_items_in_inventory: bool
    missing_items: list[str] = []


class Recipe(BaseModel):
    title: str
    theme: str | None = None
    occasion: str | None = None
    match_score: int = Field(ge=0, le=100)
    thinking_process: list[str]
    steps: list[RecipeStep]
    pro_tips: list[str] = []
    substitution_notes: list[str] = []
    estimated_time_minutes: int | None = None
    difficulty: str | None = None  # "easy" | "medium" | "hard"


class RecipeOutput(BaseModel):
    recipe: Recipe
    used_items: list[str]
    validation: RecipeValidation
